### ISSUE TITLE
Propagate generator event to all devices on tenant level (including customer entities)

### ISSUE THREAD

**Issue by lukas-k** [thingsboard/thingsboard#10128]
**Component**

 * Rule Engine

**Description**
Hi,
I have the following scenario:

In one Rule Chain, we are using a generator node to trigger some computations / updates, which need to be performed retrospectively for *all* devices inside the Tenant account. Now - how can we split the periodic generator 'trigger message' to all entities of type 'Device', including the devices owned by the customers of that Tenant?

I have tried it the following two ways:
1. Using the 'duplicate to group by name' transformation node, thus duplicating the message to the group(s) of entity type 'Device' named 'All'
![image](https://github.com/thingsboard/thingsboard/assets/13632429/7f82d02a-eb4e-4ba6-88a8-7baa3a499631)

![image](https://github.com/thingsboard/thingsboard/assets/13632429/e8e394c9-a89e-40df-a6ac-5aff7e1104da)

However, as the generator message's originator is the Tenant, it will only find devices inside the tenant's 'All' device group. Not the desired behavior.

2. Using two duplication steps: Duplicate to all the tenant's customers -> then: duplicate to all the devices of the customers device group 'All' (like above, with different originator of input message).

![image](https://github.com/thingsboard/thingsboard/assets/13632429/5a2937fd-31e0-4aa0-8346-3f0e596484fc)

The device group 'All' definitely exists inside each customer. Thus, I hoped, the second solution would work. But it does not. Still, every incoming message at the duplicate to group by name node (also with customer as originator) can not find and propagate to the Customer.All device group. It only propagates to the Tenant.All device group - thus, resulting in a number of output messages of: # customers x # tenant devices.

Does anybody see a solution for this problem, without changing the ownership structure? (bad workaround would be keeping devices inside the tenant, and only sharing access).

**Environment**

 * OS:  Ubuntu 22.04
 * ThingsBoard: 3.6.2PE
 * Browser: Firefox 123

**Comment by AndriichnekoDm** [comment]
Upon reviewing your query, we've identified it as a potential bug. Thanks for bringing this to our attention.
We're escalating this to our QA team for a detailed investigation. I'll add the fix version info as soon as it's confirmed

**Comment by AlexDoanTB** [comment]
@lukas-k It seems that you misunderstood the logic behind the rule nodes you use. 
If you duplicate to Group, the message goes to a targeted group. In your chain, you 1) send messages to all Customer entities (I assume you generate some attribute update or just trigger some action) and then 2) send the same message to other entities from the named group. Both transformation nodes are not interconnected. You have the originator that appeals to Group A and Group B afterward. It does not mean that after step 1 the message originator changes or the message goes downward through hierarchy. You named addressees and they had the package received. 
I do not see any bug. 

As your initial intent is unclear, you can use either the Relations (Create relations from devices to Customers and Duplicate to related entities with the required fetching level). Alternatively, you can utilize Duplicate to group by Name and collect devices on the particular customer level into groups with the same name (e. g. Target) and specify **Target** as the name of the group. The rule node will duplicate a message to all entities from all **Target** groups you have. If you set duplication to **All** group, then all entities from All groups within your tenant will receive the package.

**Comment by dmohns** [comment]
> As your initial intent is unclear

At least to me, the author's initial intent is clear. They want to create an event for each devices under a given tenant. Based on the screenshots provided, I assume this is used to loop all devices on a daily schedule and perform some kind of operation.

> you can use either the Relations (Create relations from devices to Customers and Duplicate to related entities with the required fetching level). 

Yes, this works, but it's a lot of effort. I think they are looking for an easier (more automated) solution that doesn't involve adding relations to every of their devices.

> Alternatively, you can utilize Duplicate to group by Name and collect devices on the particular customer level into groups with the same name (e. g. Target) and specify **Target** as the name of the group. The rule node will duplicate a message to all entities from all **Target** groups you have. 

> If you set duplication to **All** group, then all entities from All groups within your tenant will receive the package.

I don't think this is what happens. In the first scenario the event will be duplicated to only the devices in `tenant::All` group, which is most likely empty.

My assumption is that because of this behaviour the second approach was tested.

> Both transformation nodes are not interconnected. You have the originator that appeals to Group A and Group B afterward. It does not mean that after step 1 the message originator changes or the message goes downward through hierarchy. You named addressees and they had the package received. I do not see any bug.

Interesting. What would need to be done to make the second approach work, would be to add a `change originator` node between the two duplication nodes, that changes the Originator to Customer before it passes it down to the `to all devices` node.

**Comment by dmohns** [comment]
> The device group 'All' definitely exists inside each customer. Thus, I hoped, the second solution would work. But it does not. Still, every incoming message at the duplicate to group by name node (also with customer as originator) can not find and propagate to the Customer.All device group. It only propagates to the Tenant.All device group - thus, resulting in a number of output messages of: # customers x # tenant devices.

It took me a while to understand this behaviour. But taking a careful look at the *duplicate to group by name* node 

<img width="610" alt="image" src="https://github.com/thingsboard/thingsboard/assets/14202480/50ced28a-2441-4308-bd99-e054ce24cf33">

In your second node the message originator is the Customer. However, the Customers Owner is the Tenant. As a result the Group search starts at Tenant level and immediately returns with the `All` group from Tenant.

The question would be: Should it be possible to customise the *duplicate to group by name* node in a way, that the group search starts at *message originator* level rather than *message originator's owner* level?

**Comment by ashvayka** [comment]
Hi @dmohns , @lukas-k , we have merged support for this case, and it will be part of 3.6.3 release. There will be new option in the rule node configuration for this case. See below:

![image](https://github.com/thingsboard/thingsboard/assets/5016373/1804ea85-dbc9-46e0-8dba-2671c92976fd)

**Comment by dmohns** [comment]
> we have merged support for this case, and it will be part of 3.6.3 release. There will be new option in the rule node configuration for this case.

Great news! Thanks a lot. Happy to test this as soon as it's released! 

I'm not really familiar with TB release cycles yet, If I may ask, when would we be able to expect the new release 😉 ?

**Comment by ashvayka** [comment]
We plan the code freeze at the end of this week, so we expect the release in the first week of March.

**Comment by dmohns** [comment]
Hey @ashvayka 

sorry two bother I have two follow-up questions:

- Were the code changes already merged into the `master` branch (or the branch that is targeted for the `3.6.3` release)? I'm trying to have a look at the changes. Maybe even try them out from a development branch or snapshot. But I'm struggling to find them 🙈 (But maybe I'm just to unfamiliar with the TB code base organisation)
- Is there an easy way for me to follow the development of 3.6.3 and get an understand of when it might be released? Probably just the Github milestone https://github.com/thingsboard/thingsboard/milestone/11 ? Once all items are done a release will be generated?

**Comment by dmohns** [comment]
> Hey @ashvayka
> 
> sorry two bother I have two follow-up questions:
> 
> * Were the code changes already merged into the `master` branch (or the branch that is targeted for the `3.6.3` release)? I'm trying to have a look at the changes. Maybe even try them out from a development branch or snapshot. But I'm struggling to find them 🙈 (But maybe I'm just to unfamiliar with the TB code base organisation)
> * Is there an easy way for me to follow the development of 3.6.3 and get an understand of when it might be released? Probably just the Github milestone https://github.com/thingsboard/thingsboard/milestone/11 ? Once all items are done a release will be generated?

What a timing. Version *3.6.3* was released right after I wrote this message 😇 

I upgraded my instance to the latest version and tested the new `Consider originator as a group owner` feature. It works as expected.

**To solve/answer the original post for future reference:** 

To duplicate and propagate an incoming event to all devices or assets, we can combine 2 `duplicate to group by name` like so

<img width="300" alt="image" src="https://github.com/thingsboard/thingsboard/assets/14202480/f2f72a64-324b-49bd-b7e2-62aa7abbf93e">

<img width="822" alt="image" src="https://github.com/thingsboard/thingsboard/assets/14202480/9e9644bb-d64c-4cb9-9623-a5d35408faa4">

<img width="820" alt="image" src="https://github.com/thingsboard/thingsboard/assets/14202480/8101a34f-0d41-40f1-96d3-05047227ee95">

I think we can close this issue then. @lukas-k WDYT?

**Comment by lukas-k** [comment]
Sorry for the late reply. Yes, this works perfectly nice.

Thank you for the quick implementation!!
